File: java/src/main/java/com/penapereira/patterns/memento/Caretaker.java
package com.penapereira.patterns.memento;

import java.util.ArrayDeque;
import java.util.Deque;

/** Keeps mementos safe; never examines their contents (the narrow interface). */
public final class Caretaker {

    private final Deque<Originator.Memento> history = new ArrayDeque<>();

    public void save(Originator originator) {
        history.push(originator.createMemento());
    }

    /** Restores the most recent saved state; false if nothing was saved. */
    public boolean undo(Originator originator) {
        Originator.Memento last = history.poll();
        if (last == null) {
            return false;
        }
        originator.restore(last);
        return true;
    }
}

File: javascript/src/memento/caretaker.js
/** Keeps mementos safe; never examines their contents. */
export class Caretaker {
  #history = [];

  save(originator) {
    this.#history.push(originator.createMemento());
  }

  /** Restores the most recent saved state; false if nothing was saved. */
  undo(originator) {
    const last = this.#history.pop();
    if (last === undefined) {
      return false;
    }
    originator.restore(last);
    return true;
  }
}
